### PR TITLE
M2-4004, M2-4065, M2-4069, M2-4099, M2-4101, M2-4127, M2-4918, etc. : cache-related issues, app data migration infrastructure

### PR DESCRIPTION
**Pull request by anq83**
### 📝 Description

Jira tickets:
🔗 [Jira Ticket M2-4004](https://mindlogger.atlassian.net/browse/M2-4004)
🔗 [Jira Ticket M2-4069](https://mindlogger.atlassian.net/browse/M2-4069)
🔗 [Jira Ticket M2-4065](https://mindlogger.atlassian.net/browse/M2-4065)
🔗 [Jira Ticket M2-4099](https://mindlogger.atlassian.net/browse/M2-4099)
🔗 [Jira Ticket M2-4101](https://mindlogger.atlassian.net/browse/M2-4101)
🔗 [Jira Ticket M2-4127](https://mindlogger.atlassian.net/browse/M2-4127)
🔗 [Jira Ticket M2-4918](https://mindlogger.atlassian.net/browse/M2-4918)
And some other, linked.

Changes include:

- Resolves cache issues. When activity/flow is running, then user closes from the app, admin makes changes into the flow. User refreshes the app and resumes the flow. So that, the flow should use all the data-structures of the version when the flow was started.
- Adds migrations infrastructure to support the already installed and used apps and makes migration of the existing data structures to the newes format.

### CODE REVIEW

File: src/widgets/survey/model/hooks/useFlowRecordInitialization.ts [hunk 1]
```diff
@@ -45,40 +51,48 @@ export function useFlowRecordInitialization({
...
-        activityId: entityId,
+        activity: { id: activity.id, name: activity.name },
       });
     } else {
-      const activityIds = applet.activityFlows.find(
-        flow => flow.id === entityId,
-      )?.activityIds;
+      const activityIds = flow!.activityIds;
+
+      const activities = applet.activities
```
Comment (BamMironov) [inline src/widgets/survey/model/hooks/useFlowRecordInitialization.ts:64]: Please add `IdName[]` type here instead of passing it to the map function

**Review by BamMironov — CHANGES_REQUESTED** [review]

File: src/widgets/activity-group/model/hooks/useActivityGroups.ts [hunk 2]
```diff
@@ -32,9 +32,12 @@ export const useActivityGroups = (
...
     groups: groupsResult.groups,
-    isSuccess: !groupsResult.isCacheInsufficientError,
+    isSuccess:
+      !groupsResult.isCacheInsufficientError && !groupsResult.otherError,
     error: groupsResult.isCacheInsufficientError
       ? 'activity_list_component:insufficient_data_error'
+      : groupsResult.otherError
+      ? 'activity_list_component:other_error'
       : null,
   };
```
[the remark this answers is not in the document]
Comment (anq83) [reply src/widgets/activity-group/model/hooks/useActivityGroups.ts:42]: ok

File: src/widgets/survey/model/hooks/useFlowRecordInitialization.ts (hunk 1 above)
Comment (anq83) [reply src/widgets/survey/model/hooks/useFlowRecordInitialization.ts:79]: Sorrym comment is not clear

**Review by BamMironov — APPROVED** [review]

**Comment by BamMironov** [comment]
We should also create migrations for these changes.

File: src/screens/ui/Migrations/useMigration1.tsx
Comment (BamMironov) [inline src/screens/ui/Migrations/useMigration1.tsx:1]: I think we should delete this file

File: src/shared/lib/services/MigrationValidator.ts
Comment (BamMironov) [inline src/shared/lib/services/MigrationValidator.ts:1]: The file is not used either. 
Also, it has different names in code and in the file system: MigrationValidator, MigrationsValidator

File: src/shared/lib/storages/createMigrate.ts
Comment (BamMironov) [inline src/shared/lib/storages/createMigrate.ts:1]: Please delete this file too

**Review by BamMironov — CHANGES_REQUESTED** [review]

File: src/shared/lib/services/MigrationValidator.ts
Comment (anq83) [reply src/shared/lib/services/MigrationValidator.ts:1]: renamed, it's used

**Review by BamMironov — APPROVED** [review]

**Pull request merged**